File: cpp/arrays/0015-3sum.cpp
#include <algorithm>
#include <vector>

using namespace std;

vector<vector<int>> threeSum(vector<int>& nums) {
    vector<vector<int>> response;
    sort(nums.begin(), nums.end());
        
    for(int i=0; i < nums.size(); i++) {
        int n = nums[i];
        if (i > 0 && n == nums[i-1]) {
            continue;
        }
        
        int l = i + 1;
        int r = nums.size() - 1;
        while (l<r) {
            int sum = n + nums[l] + nums[r];
            if (sum > 0) {
                r--;
            }
            else if (sum < 0) {
                l++;
            }
            else {
                vector<int> match({n, nums[l++], nums[r]});
                response.push_back(match);
                while (nums[l] == nums[l-1] && l < r) {
                    l++;
                }
            }
        }
    }
    
    return response;
}

File: js/arrays/0015-3sum.js
var threeSum = function(nums) {
    nums.sort(function(a, b) {
        return a - b;
    });
    
    response = [];
    
    for (var i = 0; i < nums.length; i++) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;
        
        var l = i + 1;
        var r = nums.length - 1;
          
        while (l < r) {
            var sum = nums[i] + nums[l] + nums[r];
            
            if (sum > 0) r--;
            else if (sum < 0) l++;
            else {
                response.push([nums[i], nums[l], nums[r]]);
                l++;
                while (nums[l] == nums[l - 1] && l < r) l++;
            }
        }
    }
    
    return response;
};